### ISSUE TITLE
Investigate mitigations for "unphysical" prediction values

### ISSUE THREAD

**Issue by SteveOv** [SteveOv/ebop_maven#66]
Default model uses linear activation for all output layer predicted (regression) values.

Some values ($r_A+r_B$, $k$ and $J$) cannot <= 0, so would potentially benefit from an activation function that ensures this.

Investigate whether varying the activation function, on a value by value basis, improves the trained model.

**Comment by SteveOv** [comment]
Tried a couple of variations in the output layer, first using ReLU activation $r_A+r_B$, $k$ and $J$ and second adding $bP$ to the list. In either case the performance with the synth and formal test datasets was unconvincing.

Sample (temp) code from modelling.output_layer();
```Python
        # output_tensor = OutputLayer(units=units,
        #                             kernel_initializer=kernel_initializer,
        #                             activation=activation,
        #                             metadata=metadata,
        #                             name=name)(input_tensor)

        sumkj_output_tensor = OutputLayer(units=3,
                                         kernel_initializer=kernel_initializer,
                                         activation="relu",
                                         metadata=metadata,
                                         name=name)(input_tensor)
        poincare_output_tensor = OutputLayer(units=2,
                                           kernel_initializer=kernel_initializer,
                                           activation=activation,
                                           name="Output-Poincare")(input_tensor)
        bp_output_tensor = OutputLayer(units=1,
                                        kernel_initializer=kernel_initializer,
                                        activation="relu",
                                        name="Output-bP")(input_tensor)
        output_tensor = layers.Concatenate(axis=1, name="Concat-output")([
            sumkj_output_tensor,
            poincare_output_tensor,
            bp_output_tensor])
```

**Comment by SteveOv** [comment]
Going to look at the alternative approach of using a custom loss function to penalise predictions which are un-physical.

**Comment by SteveOv** [comment]
Here's an example of a MAE loss function which penalises implausible predictions by a factor of 10;
```Python
def mae_loss_with_penalties(y_true, y_pred):
    base_loss = losses.mean_absolute_error(y_true, y_pred)

    # All of rA_plus_rB, k, J must be > 0 and bP >= 0, so penalise implausible predictions
    implausible = (y_pred[:, 0] <= 0.) | (y_pred[:, 1] <= 0.) | (y_pred[:, 2] <= 0.)
    implausible |= y_pred[:, 5] < 0.
    return tf.where(implausible, base_loss, base_loss * 10)
```

Currently building the default model with this ...

**Comment by SteveOv** [comment]
Not great. Still getting some negative predictions with the synth-tess testing dataset.

This approach may also have some implications with training, specifically how the loss may skew the gradient descent algorithm. See [Wang et al. (2021) Understanding and Mitigating Gradient Flow Pathologies in Physics-Informed Neural Networks](https://ui.adsabs.harvard.edu/abs/2021SJSC...43A3055W)

**Comment by SteveOv** [comment]
Looking at the activation functions again, two things make it worth a revisit.
- use softplus rather than ReLU - smooth function where negative inputs are output as small positive values
- bP does not need to be forced to positive - a negative bP leads to inc > 90 which JKTEBOP is happy to work with (180 - inc)

**Comment by SteveOv** [comment]
```Python
        # Softplus for those predictions which > 0; rA+rB, k and J
        softplus_output_tensor = OutputLayer(units=3,
                                         kernel_initializer=kernel_initializer,
                                         activation="softplus",
                                         name="Output-Softplus")(input_tensor)

        # Linear for those which can be positive, zero or negative; ecosw, esinw and bP
        default_output_tensor = OutputLayer(units=3,
                                           kernel_initializer=kernel_initializer,
                                           activation=activation,
                                           metadata=metadata,
                                           name="Output")(input_tensor)
        output_tensor = layers.Concatenate(axis=1, name="Concat-Output")([
            softplus_output_tensor,
            default_output_tensor])
```

yielding
```
Metrics for 20000 system(s). - rA_plus_rB, k and J using softplus
----------------------------------------------------------------------------------------------------
Summary    | rA_plus_rB          k          J      ecosw      esinw         bP        MAE        MSE
====================================================================================================
MAE        |   0.006746   0.089082   0.063485   0.013574   0.029956   0.102979   0.050970
MSE        |   0.000112   0.046464   0.034485   0.001032   0.003068   0.050912              0.022679
```

**Comment by SteveOv** [comment]
The best results I've found are with the old distributions (k & J scale 0.4) and split output layer which gives improvements across the board when compared to the older model with all outputs using linear activation.

```
Metrics for 20000 system(s). - rA_plus_rB, k and J softplus (using new OutputLayerConcat from #67)
----------------------------------------------------------------------------------------------------
Summary    | rA_plus_rB          k          J      ecosw      esinw         bP        MAE        MSE
====================================================================================================
MAE        |   0.006927   0.069053   0.076646   0.023456   0.034658   0.065675   0.046069
MSE        |   0.000140   0.013702   0.066056   0.004221   0.004675   0.012697              0.016915
```

vs 

```
Metrics for 20000 system(s). - linear activation across whole output layer
----------------------------------------------------------------------------------------------------
Summary    | rA_plus_rB          k          J      ecosw      esinw         bP        MAE        MSE
====================================================================================================
MAE        |   0.008532   0.076823   0.087200   0.026307   0.038520   0.075263   0.052107
MSE        |   0.000231   0.016998   0.067467   0.005887   0.006263   0.016403              0.018875
```

The split model actually performs slightly worse with the real systems overall (MAE 0.056 vs 0.050) but the 20k synthetic test systems has a much broader spread and the performance there carries more weight.

**Comment by SteveOv** [comment]
Scatter of predictions for 20k synth test dataset from model with split output layer;
![predictions-nonmc-vs-labels-synthetic-mist-tess-dataset](https://github.com/user-attachments/assets/b4cdf671-73d6-4843-89d9-f2ad8ac0182d)

**Comment by SteveOv** [comment]
And the equivalent where the model with "all linear" output layer;

![predictions-nonmc-vs-labels-synthetic-mist-tess-dataset](https://github.com/user-attachments/assets/77ce63c2-77a3-42f7-ad2f-9cb00a68af67)

**Comment by SteveOv** [comment]
The mc predictions of the "split" output model for the 23 real systems;

|"split" model | "all linear" model |
| ------------- | ------------- |
| ![predictions-mc-vs-labels-model](https://github.com/user-attachments/assets/b2868a04-5975-4764-888d-2c4ab90cb0f6)  | ![predictions-mc-vs-labels-model](https://github.com/user-attachments/assets/ee20bf31-9db2-4bdb-80f4-dc390b05e34a)  |
